### PR TITLE
Fix: take platform and devices from fixtures in notify-demo tests

### LINKED ISSUE #1561: [Code Health] Four notify-demo tests hardcode platform and device IDs, ignoring --platform and --device
### Category

Robustness (potential edge-case failure)

### Component

Tests

### Description

Four `*_notify_demo` example tests are plain pytest functions whose body hardcodes both the platform string and the device IDs, so the `--platform` and `--device` options are ignored:

```python
def test_async_notify_demo() -> None:
    assert run("a2a3", [0, 1]) == 0
```

Two consequences on a shared box:

1. **Device contention.** The test grabs devices 0 and 1 whatever `task-submit` allocated. Reproduced on `upstream/main` `5bc11c2f`: three consecutive runs of `pytest examples/a2a3 tests/st/a2a3 --platform a2a3 --runtime tensormap_and_ringbuffer --device $TASK_DEVICE` on devices `1,3,5,7` gave `51 passed` / `1 failed, 50 passed` / `1 failed, 50 passed`, the failure each time being

   ```
   FAILED examples/a2a3/tensormap_and_ringbuffer/async_notify_demo/test_async_notify_demo.py::test_async_notify_demo
   E  RuntimeError: LocalMailboxEndpoint::run: child failed (worker_id=1, code=1):
      chip_process dev=1: RuntimeError: run failed with code 13
   ```

2. **Wrong-arch execution.** The a5 variant hardcodes `run("a5", ...)`, so collecting it on a2a3-only silicon builds an a5 callable and initialises an a5 worker there, and the forked chip child dies with `Fatal Python error: Segmentation fault` inside `ChipWorker.init`.

CI is not affected: it drives the scene-test scheduler, which routes `@scene_test` classes by platform, and these are plain functions it never selects. They only surface on a hand-scoped invocation — which makes them an efficient trap, because the failure looks like a regression in whatever branch is being validated. Two full hardware cycles were spent attributing them to a change before the baseline on `main` showed the same failures.

Sibling demos in the same directories already do it correctly, taking both values from fixtures:

```python
def test_sdma_async_completion_demo(st_device_ids, st_platform) -> None:
    assert run(st_platform, [int(d) for d in st_device_ids]) == 0
```

### Location

```markdown
- `examples/a2a3/tensormap_and_ringbuffer/async_notify_demo/test_async_notify_demo.py:163` — `run("a2a3", [0, 1])`
- `examples/a5/tensormap_and_ringbuffer/async_notify_demo/test_async_notify_demo.py:160` — `run("a5", [0, 1])`
- `examples/a2a3/tensormap_and_ringbuffer/deferred_notify_demo/test_deferred_notify_demo.py:175` — `run("a2a3sim", [0, 1])`
- `examples/a5/tensormap_and_ringbuffer/deferred_notify_demo/test_deferred_notify_demo.py:175` — `run("a5sim", [0, 1])`

Correct pattern, same directories:
- `examples/a2a3/tensormap_and_ringbuffer/sdma_async_completion_demo/test_sdma_async_completion_demo.py:195`
- `examples/a5/tensormap_and_ringbuffer/urma_deferred_completion_demo/test_urma_deferred_completion_demo.py:221`
```

### Proposed Fix

Take both values from the fixtures the sibling demos already use, and skip when the requested platform does not match what the demo needs:

```python
def test_async_notify_demo(st_device_ids, st_platform) -> None:
    if not st_platform.startswith("a2a3"):
        pytest.skip("a2a3 demo")
    assert run(st_platform, [int(d) for d in st_device_ids]) == 0
```

The `deferred_notify_demo` pair hardcode a *sim* platform (`a2a3sim` / `a5sim`), so they additionally need a decision on whether they are meant to be sim-only; if so, an explicit skip on a non-sim `--platform` is clearer than silently overriding it.

Worth a lint or a collection-time check that no test under `examples/` calls `run(` with a literal platform string, so the next one does not reintroduce this.

### Priority

Medium (minor risk, should fix in next few releases)

### PR DESCRIPTION
**Pull request by ChaoWao**
Fixes #1561

## Problem

Four `*_notify_demo` tests were plain pytest functions that hardcoded both the platform string and the device IDs, so `--platform` and `--device` were ignored:

```python
def test_async_notify_demo() -> None:
    assert run("a2a3", [0, 1]) == 0
```

They grabbed devices 0 and 1 whatever `task-submit` had allocated, and the a5 variants built an a5 callable on a2a3-only silicon, where the forked chip child segfaults in `ChipWorker.init`.

**Why CI never caught it.** `_collect_resource_jobs` (`conftest.py:634`) dispatches a standalone function only when it declares **both** `@pytest.mark.device_count` and `@pytest.mark.runtime`. Carrying neither, these four were collected but never selected — the full-CI invocation never ran them at all. They fired only on a hand-scoped `--runtime`/`--level` invocation, which enters child mode and bypasses the dispatcher. That is what makes them an efficient trap: the failure looks like a regression in whatever branch is being validated.

## Change

Adopt the marker + fixture pattern the sibling demos in the *same directories* already use (`sdma_async_completion_demo`, `urma_deferred_completion_demo`):

```python
@pytest.mark.platforms(["a2a3"])
@pytest.mark.runtime("tensormap_and_ringbuffer")
@pytest.mark.device_count(2)
def test_async_notify_demo(st_device_ids, st_platform) -> None:
    assert run(st_platform, [int(d) for d in st_device_ids]) == 0
```

This both honours the CLI options and enrols the tests in resource-phase dispatch, so they genuinely run in CI. `main()` and `run()` are untouched — `python test_*.py -p ... -d ...` still works.

`deferred_notify_demo` stays **sim-only** on both arches. It does pass onboard a2a3 in isolation, but it failed in the full sweep when temporarily widened, so widening it would trade one trap for another; that is logged for follow-up rather than shipped here.

## New lint

`tests/lint/check_test_platform_literals.py`, wired into pre-commit. It AST-walks test bodies under `examples/` and `tests/st/` and rejects a platform name passed as a call argument. A literal compared against the fixture (`if st_platform != "a2a3": pytest.skip(...)`) is correct usage and is not flagged, and `@pytest.mark.platforms([...])` lives in `decorator_list` so it is out of scope by construction. Platform names come from `PLATFORM_MAP` in `simpler_setup/platform_info.py`, parsed rather than imported so the hook runs in pre-commit's isolated env and cannot drift when an arch is added.

The four fixed here were the only violations in the tree.

## Verification

Onboard a2a3 (`task-submit`-locked) and sim:

- **The actual fix** — `async_notify_demo` now runs on the allocated devices: `[scheduler] START standalone test_async_notify_demo (dev=2) devices=[3, 5]` → PASS. Previously it used 0 and 1 regardless.
- **Wrong-arch trap closed** — `pytest examples/a2a3/.../{async,deferred}_notify_demo --platform a5 --collect-only` → `2 deselected` (was: collected and run).
- **Sim** — both `deferred_notify_demo` pass on `--device 6-7` under `a2a3sim` / `a5sim`, i.e. no longer pinned to 0/1.
- **No regression** — full `examples/a2a3 tests/st/a2a3 --platform a2a3 --runtime tensormap_and_ringbuffer` sweep on devices 1,3,5,7:
  - this branch: **3 failed, 50 passed, 26 deselected**
  - unmodified main @ `5768aac2`: **3 failed, 51 passed, 25 deselected**
  - **identical failure set** (`qwen3_14b_decode`, `sdma_async_completion_demo`, `prepared_callable`) — pre-existing, independent of this change. Two of the three are quarantined out of this sweep in CI (`HEAVY_IGNORE` / `SDMA_IGNORE`).

  The 51→50 / 25→26 delta is this change working: on main, `deferred_notify_demo` was selected during an *onboard* sweep and ran a `a2a3sim` workload there. It is now correctly deselected.
- `pre-commit run` — all hooks pass, including the new one; verified it fails when one file is reverted.

## Note for the a5 owner

This box is a2a3-only, so the a5 markers (`["a5"]` for async, `["a5sim"]` for deferred) preserve existing behaviour and are not onboard-validated here. Adding the markers means the a5 demos now actually run in the a5 CI job, where they previously did not — worth a look at the first a5 pipeline on this PR.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -25,6 +25,13 @@ repos:
           language_version: python3
           exclude: ^(3rdparty/|docs/zh-cn/|README\.zh-CN\.md)|vendor/
 
+        - id: check-test-platform-literals
+          name: check-test-platform-literals
+          entry: python tests/lint/check_test_platform_literals.py
+          language: python
+          language_version: python3
+          files: ^(examples|tests/st)/.*/test_[^/]*\.py$
```
[the remark this answers is not in the document]
Comment (ChaoWao) [reply .pre-commit-config.yaml]: Fixed — the file filter is now `^(examples|tests/st)/(.*/)?test_[^/]*\.py$`, so a test directly under either root (`tests/st/test_x.py`, `examples/test_x.py`) is covered while the root scope and filename pattern are unchanged.

Good catch: the checker itself already handled both shapes (`_is_checked_test_file` uses `rel.is_relative_to(root)`), so only the hook filter was narrow — the two would have silently disagreed.

Verified against the intended in/out set (old -> new):

```
tests/st/test_foo.py                            0 -> 1
examples/test_foo.py                            0 -> 1
tests/st/a5/.../test_prewarm_config.py          1 -> 1
examples/a2a3/.../test_async_notify_demo.py     1 -> 1
src/a2a3/test_x.py                              0 -> 0
tests/lint/test_x.py                            0 -> 0
```
